fix: network objects to synchronize scene changes edge cases

NoelStephensUnity opened the pull request:
The PR assures that:

- If a NetworkObject's GameObject is destroyed before it should have been, this is detected and the scene migration synchronization table is cleaned up.
- When NetworkManager shutsdown the scene migration synchronization table is cleared.

[MTTB-1025](https://jira.unity3d.com/browse/MTTB-1025)
[MTTB-1027](https://jira.unity3d.com/browse/MTTB-1027)

This PR also removes the static `NetworkObjectsToSynchronizeSceneChanges` table and migrates this script into `NetworkSpawnManager`.

fix: #3221, #3218

## Changelog

- Fixed: Issue where the scene migration synchronization table was not cleaned up if the `GameObject` of a `NetworkObject` is destroyed before it should have been.
- Fixed: Issue where the scene migration synchronization table was not cleaned up upon `NetworkManager` shutting down.

## Testing and Documentation


- Includes integration test updates to `NetworkObjectDestroyTests` and `NetworkSceneManagerEventDataPoolTest`.
- No documentation changes or additions were necessary.

<!--  Uncomment and mark items off with a * if this PR deprecates any API:
### Deprecated API
- [ ] An `[Obsolete]` attribute was added along with a `(RemovedAfter yyyy-mm-dd)` entry.
- [ ] An [api updater] was added.
- [ ] Deprecation of the API is explained in the CHANGELOG.
- [ ] The users can understand why this API was removed and what they should use instead.
-->

NoelStephensUnity committed:
fix

Clean up scene migration synchronization table entry when GameObject is destroyed when it should not have been.

NoelStephensUnity committed:
fix

Clean up scene migration synchronization table when NetworkManager shuts down.

NoelStephensUnity committed:
fix

adding check for any in-scene placed NetworkObjects.
removing whenever destroyed (if entry still exists).

NoelStephensUnity committed:
test

Minor test to validate the NetworkObject is no longer within the NetworkObject.NetworkObjectsToSynchronizeSceneChanges list upon destroying it prior to it being despawned.

NoelStephensUnity committed:
update

Adding change log entries.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/NetworkObjectsToSynchronizeSceneChanges-edge-cases

NoelStephensUnity committed:
fix

Removing static NetworkObjectsToSynchronizeSceneChanges and CleanUpDisposedObjects and migrating them to normal properties within the NetworkSpawnManager.

NoelStephensUnity committed:
test

Adding a validation at the end of the scene load-unload test to assure the NetworkObjectsToSynchronizeSceneChanges has no remaining in-scene placed NetworkObjects when all scenes are unloaded.
Removing a verbose setting from InScenePlacedNetworkObjectTests.

NoelStephensUnity committed:
fix

Fixing issue with logic for scene migration synchronization.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -1196,6 +1187,50 @@ private void SpawnNetworkObjectLocallyCommon(NetworkObject networkObject, ulong
+            {
+                NetworkObjectsToSynchronizeSceneChanges.Remove(networkObjectId);
+            }
+            CleanUpDisposedObjects.Clear();
```
It looks like `CleanUpDisposedObjects` is only added to from within this function, and then is cleared within this function. Should the list either only exist here, or the two loops be collapsed?

NoelStephensUnity replied:
Changed how this is done.

NoelStephensUnity committed:
update

switching from a list to a stack

NoelStephensUnity committed:
test

Had to adjust for the changes from static to NetworkManager relative (for DA mode).

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/NetworkObjectsToSynchronizeSceneChanges-edge-cases
